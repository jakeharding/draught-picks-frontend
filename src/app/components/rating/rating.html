<!-- Generated template for the RatingComponent component -->
<ion-row>
  <div (click)="updateRating(1)" [ngClass]="{'color-mug': rating?.rating >= 1, 'no-color-mug': rating?.rating < 1, 'large': size == 'large'}"></div>
  <div (click)="updateRating(2)" [ngClass]="{'color-mug': rating?.rating >= 2, 'no-color-mug': rating?.rating < 2, 'large': size == 'large'}"></div>
  <div (click)="updateRating(3)" [ngClass]="{'color-mug': rating?.rating >= 3, 'no-color-mug': rating?.rating < 3, 'large': size == 'large'}"></div>
  <div (click)="updateRating(4)" [ngClass]="{'color-mug': rating?.rating >= 4, 'no-color-mug': rating?.rating < 4, 'large': size == 'large'}"></div>
  <div (click)="updateRating(5)" [ngClass]="{'color-mug': rating?.rating >= 5, 'no-color-mug': rating?.rating < 5, 'large': size == 'large'}"></div>
</ion-row>
