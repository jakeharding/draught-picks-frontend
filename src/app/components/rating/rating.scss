:host {
  display: flex;
  flex-direction: row;

  .color-mug{
    background-image: url("../../../assets/imgs/rating/mug_color.png");
    background-size: cover;
    height: 30px;
    width: 30px;
    margin-right: 5px;
  }

  .no-color-mug {
    background-image: url("../../../assets/imgs/rating/mug.png");
    background-size: cover;
    height: 30px;
    width: 30px;
    margin-right: 5px;
  }

  div {
    flex: 1;
  }

  .large {
    height: 55px;
    width: 55px;
    margin-right: 10px;
  }

  @media (max-width: 576px) {
    .large {
      height: 50px;
      width: 50px;
      margin-right: 0px;
    }
  }
}
