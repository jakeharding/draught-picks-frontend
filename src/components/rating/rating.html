<!-- Generated template for the RatingComponent component -->
<ion-row>
  <div [ngClass]="{'color-mug': rating?.rating >= 1, 'no-color-mug': rating?.rating < 1, 'large': size == 'large'}"></div>
  <div [ngClass]="{'color-mug': rating?.rating >= 2, 'no-color-mug': rating?.rating < 2, 'large': size == 'large'}"></div>
  <div [ngClass]="{'color-mug': rating?.rating >= 3, 'no-color-mug': rating?.rating < 3, 'large': size == 'large'}"></div>
  <div [ngClass]="{'color-mug': rating?.rating >= 4, 'no-color-mug': rating?.rating < 4, 'large': size == 'large'}"></div>
  <div [ngClass]="{'color-mug': rating?.rating >= 5, 'no-color-mug': rating?.rating < 5, 'large': size == 'large'}"></div>
</ion-row>
