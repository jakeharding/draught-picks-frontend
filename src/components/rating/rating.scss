rating {
  .color-mug{
    background-image: url("../../assets/imgs/rating/mug_color.png");
    background-size: cover;
    height: 30px;
    width: 30px;
    margin-right: 5px;
  }

  .no-color-mug {
    background-image: url("../../assets/imgs/rating/mug.png");
    background-size: cover;
    height: 30px;
    width: 30px;
    margin-right: 5px;
  }
}
