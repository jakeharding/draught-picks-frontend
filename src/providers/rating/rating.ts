import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import Env from "../../env";
import BeerRating from "../../models/BeerRating";

/**
 * Generated class for the RatingProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
**/
@Injectable()
export class RatingProvider {

  url: string;

  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/beer-ratings`;
  }

  /**
   * create function
   * Parameters: beerRating
   * Creates a beer rating for a user
   * */
  create(beerRating: BeerRating): Promise<BeerRating> {
    return this.http.post(this.url, beerRating).toPromise().then((rating:BeerRating) => rating);
  }

  /**
   * partialUpdate function
   * Parameters: beerRating
   * Updates beer rating
   * */
  partialUpdate(beerRating: BeerRating): Promise<BeerRating> {
    return this.http.patch(
      `${this.url}/${beerRating.uuid}`, beerRating).toPromise().then((rating:BeerRating) => rating);
  }
}
