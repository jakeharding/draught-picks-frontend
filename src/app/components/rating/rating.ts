import { Component, EventEmitter, Input, Output } from '@angular/core';
import BeerRating from '../../models/BeerRating';
import { RatingProvider } from '../../services/rating/rating';
import { ToastController } from '@ionic/angular';

/**
 * Generated class for the RatingComponent component.
 *
 * See https://angular.io/api/core/Component for more info on Angular
 * Components.
 */
@Component({
  selector: 'rating',
  templateUrl: 'rating.html',
  styleUrls: ['./rating.scss']
})
export class RatingComponent {

  constructor(public ratingProvider: RatingProvider, public toastController: ToastController) {}

  private static readonly LARGE = 'large';

  @Input() rating: BeerRating = { beer: 'beerId'} as BeerRating;
  @Input() size = 'small';
  @Output() setRating = new EventEmitter<BeerRating>();

  /**
   * updateRating function
   * Parameters: rating
   * updates the rating of a beer
   */
  updateRating(rating: number) {
    if (this.size === RatingComponent.LARGE) {
      const success = async beerRating => {
        this.rating = beerRating; // Update the beerRating returned from REST API
        const toast = await this.toastController.create({
          message: 'Your rating is saved.',
          duration: 3000,
          position: 'top',
          cssClass: 'success-toast'
        });
        await toast.present();
        this.setRating.emit(beerRating); // Update parent component
      };

      if (this.rating && this.rating.uuid) {
        return this.ratingProvider.partialUpdate({rating, uuid: this.rating.uuid} as BeerRating).then(success);
      } else {
        return this.createRating(this.rating.beer, rating).then(success);
      }
    }
  }

  /**
   * createRating function
   * Parameters: rating
   * creates a new rating for a beer if none exist
   */
  createRating(beer: string, rating: number) {
    const newRating = {beer, rating} as BeerRating;
    return this.ratingProvider.create(newRating);
  }
}
