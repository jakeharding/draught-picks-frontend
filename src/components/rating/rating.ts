import {Component, EventEmitter, Input, Output} from '@angular/core';
import BeerRating from "../../models/BeerRating";
import {RatingProvider} from "../../providers/rating/rating";
import {ToastController} from "ionic-angular";

/**
 * Generated class for the RatingComponent component.
 *
 * See https://angular.io/api/core/Component for more info on Angular
 * Components.
 */
@Component({
  selector: 'rating',
  templateUrl: 'rating.html'
})
export class RatingComponent {

  @Input() rating: BeerRating;
  @Input() size: string = 'small';
  @Output() setRating = new EventEmitter<BeerRating>();

  private static LARGE = 'large';

  constructor(public ratingProvider: RatingProvider, public toastController: ToastController) {}

  /**
   * updateRating function
   * Parameters: rating
   * updates the rating of a beer
   * */
  updateRating (rating: number) {
    if (this.size === RatingComponent.LARGE) {
      const success = rating => {
        this.rating = rating;
        const toast = this.toastController.create({
          message: "Your rating is saved.",
          duration: 3000,
          position: "top",
          cssClass: "success-toast"
        });
        toast.present();
        this.setRating.emit(rating); // Update parent component
      };

      if (this.rating && this.rating.uuid) {
        return this.ratingProvider.partialUpdate({rating, uuid: this.rating.uuid} as BeerRating).then(success)
      } else {
        this.createRating(this.rating.beer, rating).then(success);
      }
    }
  }

  /**
   * createRating function
   * Parameters: rating
   * creates a new rating for a beer if none exist
   * */
  createRating (beer: string, rating: number) {
    const newRating = {beer, rating} as BeerRating;
    return this.ratingProvider.create(newRating);
  }
}
