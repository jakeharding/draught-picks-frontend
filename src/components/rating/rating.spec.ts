import { async, ComponentFixture, TestBed } from "@angular/core/testing";
import { RatingComponent } from "./rating";
import { CUSTOM_ELEMENTS_SCHEMA } from "@angular/core";
import { ToastController } from "ionic-angular";
import { RatingProvider } from "../../providers/rating/rating";

/**
 * rating.spec.ts
 *
 * Created by jake
 * Created on 11/15/18
 *
 * Test the RatingComponent.
 */

describe('RatingComponent', () => {
  let component: RatingComponent;
  let fixture: ComponentFixture<RatingComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [RatingComponent],
      providers: [
        {provide: RatingProvider, useValue: {}},
        {provide: ToastController, useValue: {}}
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(RatingComponent);
    component = fixture.debugElement.componentInstance;
    // jest.clearAllMocks();
  });

  it('should initialize the RatingComponent', () => {
    expect(component).toBeDefined();
    expect(fixture).toMatchSnapshot();
  });


});
