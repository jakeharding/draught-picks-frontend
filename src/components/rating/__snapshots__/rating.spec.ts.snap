// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingComponent should initialize the RatingComponent 1`] = `
<rating
  ratingProvider={[Function Object]}
  setRating={[Function EventEmitter]}
  size={[Function String]}
  toastController={[Function Object]}
>
  
  <ion-row>
    
    
    <div />
    
    
    <div />
    
    
    <div />
    
    
    <div />
    
    
    <div />
    
  
  </ion-row>
  
</rating>
`;
